model/src/main/kotlin/utils/PostgresFileArchiverStorage.kt

```diff
@@ -0,0 +1,144 @@
+import org.jetbrains.exposed.sql.SchemaUtils.createMissingTablesAndColumns
+import org.jetbrains.exposed.sql.SchemaUtils.withDataBaseLock
+import org.jetbrains.exposed.sql.transactions.transaction
+
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.checkDatabaseEncoding
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.tableExists
+import org.ossreviewtoolkit.utils.ORT_NAME
+import org.ossreviewtoolkit.utils.log
+
+/**
+ * A PostgreSQL based storage for archive files.
+ */
+class PostgresFileArchiverStorage(
+    /**
+     * The JDBC data source to obtain database connections.
+     */
+    dataSource: DataSource
+) : FileArchiverStorage {
+    init {
+        Database.connect(dataSource).apply { defaultFetchSize(1000) }
+
+        transaction {
+            withDataBaseLock {
+                if (!tableExists(FileArchiveTable.tableName)) {
+                    checkDatabaseEncoding()
+                    createMissingTablesAndColumns(FileArchiveTable)
+                }
+            }
+
+            commit()
+        }
+    }
+
+    override fun hasArchive(provenance: Provenance): Boolean {
+        checkIsUniqueStorageKey(provenance)
+
+        return transaction {
+            queryFileArchive(provenance)
+        } != null
+    }
+
+    override fun addArchive(provenance: Provenance, zipFile: File) {
+        checkIsUniqueStorageKey(provenance)
+
+        transaction {
+            if (queryFileArchive(provenance) == null) {
+                try {
+                    FileArchive.new {
+                        this.provenance = provenance.storageKey()
+                        this.zipData = zipFile.readBytes()
+                    }
+                } catch (e: ExposedSQLException) {
+                    // The exception can happen when an archive with the same provenance has been inserted in parallel.
+                    // That race condition is possible because [java.sql.Connection.TRANSACTION_READ_COMMITTED] is used
+                    // as transaction isolation level (by default).
+                    log.warn(e) { "Could not insert archive for '${provenance.storageKey()}'." }
+                }
+            }
+        }
+    }
+
+    override fun getArchive(provenance: Provenance): File? {
+        checkIsUniqueStorageKey(provenance)
+
+        val fileArchive = transaction {
+            queryFileArchive(provenance)
+        } ?: return null
+
+        val file = createTempFile(ORT_NAME, ".zip").toFile()
+
+        try {
+            file.writeBytes(fileArchive.zipData)
+        } catch (e: IOException) {
+            file.delete()
+            throw e
+        }
+
+        return file
+    }
+}
+
+private object FileArchiveTable : IntIdTable("file_archives") {
+    val provenance: Column<String> = text("provenance")
+    val zipData: Column<ByteArray> = binary("zip_data")
+
+    init {
+        uniqueIndex(provenance)
+    }
```

Comment:
You could use `text("provenance").uniqueIndex()` above to make it a bit shorter.